timApp/tide_support/utils.py

```diff
@@ -0,0 +1,504 @@
+    """
+    Header of the plugin
+    """
+
+    stem: str | None = None
+    """
+    Stem of the plugin
+    """
+
+    answer_count: int | None = None
+    """
+    Number of answers for the task
+    """
+
+    # Might be needed to require this
+    type: str | None = None
+    """ Type of the file """
+
+
+TIDETaskInfoSchema = class_schema(TIDETaskInfo)()
+
+
+@dataclass
+class TIDEPluginData:
+    """
+    Data for the TIDE-task
+    """
+
+    task_files: List[IdeFile] | None = None
+    """
+    Files for the TIDE-task
+    """
+
+    header: str | None = None
+    """
+    Header of the plugin
+    """
+
+    stem: str | None = None
+    """
+    Stem of the plugin
+    """
+
+    type: str | None = None
+    """
+    Type of the file
+    """
+
+    task_id: str | None = None
+    """
+    Task id
+    """
+
+    doc_id: int | None = None
+    """
+    Document id
+    """
+
+    par_id: int | None = None
+    """
+    Paragraph id
+    """
+
+    ide_task_id: str | None = None
+    """
+    TIDE-task id
+    """
+
+
+IDE_TASK_TAG = "ideTask"  # Identification tag for the TIDE-task
+
+
+@dataclass
+class TIDEError:
+    """
+    Error message for the TIDE-task
+    """
+
+    error: str
+
+
+def user_ide_courses(user: User) -> list[json] | TIDEError:
+    """
+    Gets all courses that have parameter for Ide course in course settings and are bookmarked by the user
+    :param user: Logged-in user
+    :return: List JSON with all TIDE-tasks from the courses user has bookmarked
+    """
+
+    if not user.bookmarks.bookmark_data and len(user.bookmarks.bookmark_data) < 2:
+        raise NotExist()
+
+    user_courses = user.bookmarks.bookmark_data[2].get(MY_COURSES_GROUP)
+
+    # Add hidden courses to the list
+    if user.bookmarks.bookmark_data[2].get(HIDDEN_COURSES_GROUP):
+        user_courses.extend(user.bookmarks.bookmark_data[2][HIDDEN_COURSES_GROUP])
+
+    if not user_courses:
+        return TIDEError("No courses found")
+
+    ide_courses = []
+
+    for course_dict in user_courses:
+        for course_name, course_path in course_dict.items():
+            # Get the document by path, remove /view/ from the beginning of path
+            doc = DocInfo.find_by_path(course_path.split("view/")[1:][0])
+
+            if doc is None:
+                continue
+
+            task_paths = doc.document.get_settings().ide_course()
+            paths = []
+            for path in task_paths:
+                paths.append(path.path)
+
+            if task_paths is None:
+                continue
+
+            ide_courses.append(
+                {
+                    "name": course_name,
+                    "id": doc.document.doc_id,
+                    "path": course_path,
+                    "task_paths": paths,
+                }
+            )
+
+    if not ide_courses:
+        return TIDEError("No courses found")
+
+    return ide_courses
+
+
+def ide_task_folders_by_doc(
+        doc_id: int = None, doc_path: str = None
```

Comment:
Tyypitys: None pitää merkitä tyyppiin erikseen, esim `int | None`.

```suggestion
def ide_task_folders_by_doc(
        doc_id: int | None = None, doc_path: str | None = None
```

Reply:
Korjattu